src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1185,143 @@ void Client::HandleTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+    mTcpInitialized = false;
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
+}
+
+void Client::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpEstablished(aEndpoint);
+}
+
+void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                       size_t         aBytesAvailable,
+                                       bool           aEndOfStream,
+                                       size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message              *message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+    uint16_t              length = 0;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        mTcpTimer.Stop();
+        mEndpoint.SendEndOfStream();
+    }
+
+    VerifyOrExit(aBytesAvailable < 2);
+    SuccessOrExit(mEndpoint.ReceiveByReference(data));
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    if (data != nullptr)
+    {
+        totalReceived += data->mLength;
+
+        if (data->mLength == 1)
+        {
+            length |= (uint16_t)data->mData[0] << 8;
+            if ((data = data->mNext) != nullptr)
+            {
+                length |= (uint16_t)data->mData[0] & 0xff;
+                length = HostSwap16(length);
+                totalReceived += data->mLength;
+                SuccessOrExit(message->AppendBytes(data->mData + 1, data->mLength - 1));
+            }
+        }
+        else if (data->mLength >= 2)
+        {
+            length = ReadUint16(data->mData);
+            SuccessOrExit(message->AppendBytes(data->mData + sizeof(uint16_t), data->mLength - sizeof(uint16_t)));
+        }
+
+        // Check if whole query resonse was received.
+        VerifyOrExit((length + sizeof(uint16)) > aBytesAvailable);
+
+        for (data = data->mNext; data != nullptr; data = data->mNext)
+        {
+            SuccessOrExit(message->AppendBytes(data->mData, data->mLength));
+            totalReceived += data->mLength;
+        }
+    }
+
+    mEndpoint.CommitReceive(totalReceived, 0);
+    response.mInstance = &Get<Instance>();
+    response.mMessage  = message;
+
+    SuccessOrExit(ParseResponse(response, type, responseError));
+
+    if (responseError == kErrorNone && length > OPENTHREAD_CONFIG_DNS_OVER_TCP_QUERY_MAX_SIZE)
+    {
+        LogWarn("Dns query over wasn't received message too big.");
+        responseError = kErrorNoBufs;
+    }
+
+    FinalizeQuery(response, type, responseError);
```

Comment:
@canisLupus1313  I think we can simplfiy the processing if we add a helper method to read and copy content from a `LinkedBuffer` to a `Message`:

```c++
Error Client::ReadFromLinkBuffer(const otLinkedBuffer *&aLinkedBuffer,
                                 size_t                &aOffset,
                                 Message               &aMessage,
                                 uint16_t               aLength)
{
    // Read `aLength` bytes from `aLinkedBuffer` starting at `aOffset`
    // and copy the content into `aMessage`. As we read we can move
    // to the next `aLinkedBuffer` and update `aOffset`.
    // Returns:
    // - `kErrorNone` if `aLengh` bytes are successfully read and
    //    `aOffset` and `aLinkedBuffer` are updated.
    // - `kErrorNotFound` is not enough bytes available to read
    //    from `aLinkedBuffer`.
    // - 'kErrorNotBufs` if cannot grow `aMessage` to append bytes.

    Error error = kErrorNone;

    while (aLength > 0)
    {
        uint16_t bytesToRead = aLength;

        VerifyOrExit(aLinkedBuffer != nullptr, error = kErrorNotFound);

        if (bytesToRead > aLinkedBuffer->mLength - aOffset)
        {
            bytesToRead = static_cast<uint16_t>(aLinkedBuffer->mLength - aOffset);
        }

        SuccessOrExit(error = aMessage.AppendBytes(&aLinkedBuffer->mData[aOffset], bytesToRead));

        aLength -= bytesToRead;
        aOffset += bytesToRead;

        if (aOffset == aLinkedBuffer->mLength)
        {
            aLinkedBuffer = aLinkedBuffer->mNext;
            aOffset       = 0;
        }
    }

exit:
    return error;
}
```

Reply:
We can then use it to process the received TCP buffer. We can reuse the same `message` instance and handle the case if we happen to get multiple response in TCP rx buffer. Or if there are leftover bytes at the end (we call `CommitReceive()` on bytes we read and process). Here is my attempt at writing this (feel free to use it or change it) 😃 
```c++
void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
                                       size_t         aBytesAvailable,
                                       bool           aEndOfStream,
                                       size_t         aBytesRemaining)
{
    OT_UNUSED_VARIABLE(aEndpoint);
    OT_UNUSED_VARIABLE(aBytesRemaining);

    Message              *message   = nullptr;
    size_t                totalRead = 0;
    size_t                offset    = 0;
    const otLinkedBuffer *data;

    if (aEndOfStream)
    {
        // Cleanup is done in dissconnected callback.
        mTcpTimer.Stop();
        IgnoreError(mEndpoint.SendEndOfStream());
    }

    SuccessOrExit(mEndpoint.ReceiveByReference(data));
    VerifyOrExit(data != nullptr);

    message = mSocket.NewMessage(0);
    VerifyOrExit(message != nullptr);

    while (aBytesAvailable > totalRead)
    {
        uint16_t length;

        // Read the 'length' field.
        SuccessOrExit(ReadFromLinkBuffer(data, offset, *message, sizeof(uint16_t)));

        IgnoreError(message->Read(/* aOffset */ 0, length));
        length = HostSwap16(length);

        // Try to read `length` bytes.
        IgnoreError(message->SetLength(0));
        SuccessOrExit(ReadFromLinkBuffer(data, offset, *message, length));

        totalRead += length + sizeof(uint16_t);

        // Now process the read message as query response.
        {
            Response  response;
            QueryType type;
            Error     responseError;

            response.mInstance = &Get<Instance>();
            response.mMessage  = message;

            if (ParseResponse(response, type, responseError) == kErrorNone)
            {
                if (responseError == kErrorNone && length > OPENTHREAD_CONFIG_DNS_OVER_TCP_QUERY_MAX_SIZE)
                {
                    LogWarn("Dns query over wasn't received message too big.");
                    responseError = kErrorNoBufs;
                }

                FinalizeQuery(response, type, responseError);
            }
        }

        IgnoreError(message->SetLength(0));

        // Loop again to see if we can read another response.
    }

exit:
    // Inform `mEndPoint` about the total read and processed bytes
    IgnoreError(mEndpoint.CommitReceive(totalRead, /* aFlags */ 0));
    FreeMessage(message);
}
```

Reply:
@abtink Thanks I have applied this change looks like this implementation works.